### PR TITLE
AB#18 AB#19 AB#20 - PR 

### PR DESCRIPTION
**Pull request by minhman97**
[AB#18](https://dev.azure.com/minhmancarter/7844c8b2-5ce7-4cec-baf2-5a3248015162/_workitems/edit/18) [AB#19](https://dev.azure.com/minhmancarter/7844c8b2-5ce7-4cec-baf2-5a3248015162/_workitems/edit/19) [AB#20](https://dev.azure.com/minhmancarter/7844c8b2-5ce7-4cec-baf2-5a3248015162/_workitems/edit/20)

### CODE REVIEW

File: BackEnd/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# syntax=docker/dockerfile:1
+
+FROM mcr.microsoft.com/dotnet/sdk:7.0 as build-env
+
+WORKDIR /source
+
+COPY . .
+
+RUN dotnet restore --disable-parallel
```
Comment (popemkt) [inline BackEnd/Dockerfile]: Remove --disable-parallel

File: BackEnd/SnippetManagement.Api/Program.cs [hunk 2]
```diff
@@ -44,7 +46,9 @@
     opts.AddPolicy(name: "_myAllowSpecificOrigins",
         policy  =>
         {
-            policy.WithOrigins("http://localhost:3000").AllowAnyHeader().AllowAnyMethod();
+            policy.WithOrigins("http://localhost:5002", "http://localhost:3000")
```
Comment (popemkt) [inline BackEnd/SnippetManagement.Api/Program.cs]: No debug values
value 5002 must comes from 1 source of truth

File: BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs [hunk 3]
```diff
@@ -26,20 +29,41 @@ public AuthenticationService(IOptions<JwtConfiguration> jwtConfiguration, IUnitO
...
+                    SocialProvider = SocialProvider.Google
+                });
+            }
             SecurityTokenDescriptor tokenDescriptor = GetTokenDescriptor(user);
             var tokenHandler = new JwtSecurityTokenHandler();
             SecurityToken securityToken = tokenHandler.CreateToken(tokenDescriptor);
-            string token = tokenHandler.WriteToken(securityToken);
-
-            return token;
+            return Result.Ok().WithSuccess(tokenHandler.WriteToken(securityToken));
```
Comment (popemkt) [inline BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs]: Incorrect use of Result Class

File: FrontEnd/snippet_management.web/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# build environment
+FROM node:alpine as build
+WORKDIR /app
+
+COPY package*.json .
+RUN npm install --silent --force
+COPY . .
+
+RUN npm run build -- --profile
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/Dockerfile]: CRA already supports production build with build cmd

File: FrontEnd/snippet_management.web/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# build environment
+FROM node:alpine as build
+WORKDIR /app
+
+COPY package*.json .
+RUN npm install --silent --force
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/Dockerfile]: no need for silent

File: FrontEnd/snippet_management.web/src/api/ApiEndpoint.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export const baseUrl = process.env.REACT_APP_BASE_URL;
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/src/api/ApiEndpoint.js:1]: Should centralize all urls here

File: FrontEnd/snippet_management.web/src/api/ProgramingLanguageApi.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { baseUrl } from "./ApiEndpoint";
+import { GetErrorMessage } from "./StatusCode";
+import { getUserToken } from "./UserApi";
+
+export const getProgramingLanguages = async () => {
+  const token = getUserToken();
+  return await fetch(`${baseUrl}ProgramingLanguage/GetLanguage`, {
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/src/api/ProgramingLanguageApi.js]: Fix this endpoint naming => REST

File: FrontEnd/snippet_management.web/src/api/UserApi.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    return userToken?.token;
+  }
+
+  return undefined;
+};
+
+export const removeToken = () => {
+  window.location = "/";
+  localStorage.removeItem(keyToken);
+};
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/src/api/UserApi.js]: Doesn't belong to API

File: FrontEnd/snippet_management.web/src/api/StatusCode.js [hunk 9]
```diff
@@ -1,3 +1,5 @@
+import { removeToken } from "./UserApi";
+
 export const GetErrorMessage = (statusCode) => {
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/src/api/StatusCode.js]: Need more suitable name

File: FrontEnd/snippet_management.web/src/components/LoginForm.js [hunk 10]
```diff
@@ -1,23 +1,67 @@
-import { React, useState } from "react";
+import { React, useEffect, useState } from "react";
 import { login } from "../api/UserApi";
 
 const LoginForm = ({ setToken }) => {
   const [email, setEmail] = useState(null);
   const [password, setPassword] = useState(null);
+  const [Auth, setAuth] = useState({ isAuth: false, message: "" });
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/src/components/LoginForm.js]: Auth no Cap
Should use null to indicate no value

File: FrontEnd/snippet_management.web/src/routes/Snippet/Snippet.js [hunk 11]
```diff
@@ -21,19 +20,17 @@ const Snippet = () => {
 
   const { sortOrder, setSortOrder } = usePaginationContext();
 
-  const [token] = useToken();
-
-  const { mutate: muateDeleteSnippet } = useDeleteSnippet();
-  const { mutate: muateUpdateSnippet } = useUpdateSnippet();
+  const { mutate: muatateDeleteSnippet } = useDeleteSnippet();
```
Comment (popemkt) [inline FrontEnd/snippet_management.web/src/routes/Snippet/Snippet.js]: fix typo

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+      Jwt__ExpiringDays: 1
+      ASPNETCORE_URLS: http://*:${API_PORT}
+  db:
+    image: mcr.microsoft.com/mssql/server:2019-latest
+    volumes:
+      - snippet-management-vol:/var/opt/mssql
+    ports:
+      - '5000:1433'
+    environment:
+      MSSQL_SA_PASSWORD: MyPassword123!
```
Comment (popemkt) [inline docker-compose.yml]: Should be variable

File: docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+      ACCEPT_EULA: 'Y'
+  web:
+    image: snippet-management-web
+    ports:
+      - '${WEB_PORT}:3000'
+    environment:
+      REACT_APP_GOOGLE_CLIENT_ID: 448086775090-g62vpn8qbig7gqt2qiin9m6ok5sgislk
```
Comment (popemkt) [inline docker-compose.yml]: Should be variable

File: docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    image: snippet-management
+    ports:
+      - '${API_PORT}:${API_PORT}'
+    depends_on:
+      - db
+    environment:
+      ConnectionStrings__DefaultConnection: Server=db;Database=SnippetManagement;User Id=sa;Password=MyPassword123!;TrustServerCertificate=true
+      Jwt__ValidAudience: http://localhost:${API_PORT}
+      Jwt__ValidIssuer: http://localhost:${API_PORT}
+      Jwt__IssuerSigningKey: ByYM000OLlMQG6VVVp1OH7Xzyr7gHuw1qvUC5dcGt3SNM
```
Comment (popemkt) [inline docker-compose.yml]: Should be variable

**Review by popemkt — DISMISSED** [review]

File: BackEnd/SnippetManagement.Api/Controllers/AuthenticationController.cs [hunk 15]
```diff
@@ -19,18 +19,29 @@ public AuthenticationController(IAuthenticationService authenticationService)
...
             Email = request.Email,
             Password = request.Password
-        });
-        if (string.IsNullOrEmpty(token))
+        }, string.Empty);
+        if (result.IsFailed)
         {
-            ModelState.AddModelError(string.Empty, "Invalid login");
-            return BadRequest(ModelState);
+            return BadRequest(result);
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Controllers/AuthenticationController.cs]: Fix this

File: BackEnd/SnippetManagement.Api/Controllers/SnippetController.cs [hunk 16]
```diff
@@ -84,8 +88,8 @@ public async Task<IActionResult> GetSnippets(int startIndex, int endIndex, [From
         [FromQuery] SortOrder sortOrder)
     {
         var userId = User.Claims.Where(c => c.Type == "UserId")
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Controllers/SnippetController.cs]: Refactor this into IdentityService.GetCurrentUserId()

File: docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+      - db
+    environment:
+      ConnectionStrings__DefaultConnection: Server=db;Database=SnippetManagement;User Id=sa;Password=${MSSQL_SA_PASSWORD};TrustServerCertificate=true
+      Jwt__ValidAudience: http://localhost:${API_PORT}
+      Jwt__ValidIssuer: http://localhost:${API_PORT}
+      Jwt__IssuerSigningKey: ${JWT_ISSUER_SIGNING_KEY}
+      Jwt__ExpiringDays: 1
+      ASPNETCORE_URLS: http://*:${API_PORT}
+      API_PORT: ${API_PORT}
+      WEB_PORT: ${WEB_PORT}
```
Comment (minhman97) [inline docker-compose.yml]: Rename following .net config convention

File: FrontEnd/snippet_management.web/src/api/ApiEndpoint.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+export const authUrl = `${baseUrl}/Authentication`;
+export const externalAuthUrl = `${authUrl}/External`;
+
+export const programingLanguageUrl = `${baseUrl}/ProgramingLanguage`;
+
+export const getSnippetUrl = (startIndex, endIndex, sortProperty, orderWay) => {
+  return `${baseUrl}/Snippet?startIndex=${startIndex}&endIndex=${endIndex}&property=${sortProperty}&orderWay=${orderWay}`;
+};
+
+export const searchSnippetUrl = (
```
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/api/ApiEndpoint.js]: merge into 1 method

File: FrontEnd/snippet_management.web/src/components/SnippetList.js [hunk 19]
```diff
@@ -81,7 +81,7 @@ const SnippetList = ({ pages, fetchNextPage, handleUpdateSnippet }) => {
                             alt="programing language name"
                             src={`${
                               baseUrl +
```
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/components/SnippetList.js]: Move to correct location

File: FrontEnd/snippet_management.web/src/hooks/SnippetHooks.js
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/hooks/SnippetHooks.js:1]: Rename

File: FrontEnd/snippet_management.web/src/routes/Snippet/CreateSnippet.js [hunk 20]
```diff
@@ -69,7 +68,7 @@ const CreateSnippet = () => {
                     <div className="flex items-center">
                       <img
                         alt="test"
-                        src={`${baseUrl + language.url}`}
+                        src={`${baseUrl}/${language.url}`}
```
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/routes/Snippet/CreateSnippet.js]: move to correct location

File: docker-compose.yml [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+      - snippet-management-vol:/var/opt/mssql
+    ports:
+      - '5000:1433'
+    environment:
+      MSSQL_SA_PASSWORD: ${MSSQL_SA_PASSWORD}
+      ACCEPT_EULA: 'Y'
+  web:
+    image: snippet-management-web
+    ports:
+      - '${WEB_PORT}:3000'
```
Comment (minhman97) [inline docker-compose.yml]: Make this variable as well

File: docker-compose.yml [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+      Jwt__ExpiringDays: 1
+      ASPNETCORE_URLS: http://*:${API_PORT}
+      API_PORT: ${API_PORT}
+      WEB_PORT: ${WEB_PORT}
+  db:
+    image: mcr.microsoft.com/mssql/server:2019-latest
+    volumes:
+      - snippet-management-vol:/var/opt/mssql
+    ports:
+      - '5000:1433'
```
Comment (minhman97) [inline docker-compose.yml]: Make this variable as well

File: BackEnd/SnippetManagement.Api/Controllers/AuthenticationController.cs [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+    public async Task<IActionResult> GetAuthToken(LoginRequest request)
+    {
+        var result = await _authenticationService.GetToken(new UserDto()
+        {
+            Email = request.Email,
+            Password = request.Password
+        }, string.Empty);
+        if (result.IsFailed)
+        {
+            return BadRequest(new { isFailed = result.IsFailed, message = result.Reasons.FirstOrDefault() });
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Controllers/AuthenticationController.cs:29]: No need for isFailed status

File: BackEnd/SnippetManagement.Api/Controllers/SnippetController.cs [hunk 24]
```diff
@@ -14,10 +15,12 @@ namespace SnippetManagement.Api.Controllers;
 public class SnippetController : ControllerBase
 {
     private readonly IUnitOfWork _unitOfWork;
+    private IIdentityService identityService;
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Controllers/SnippetController.cs:18]: Fix this

File: BackEnd/SnippetManagement.Api/Program.cs [hunk 25]
```diff
@@ -39,30 +42,31 @@
 });
 builder.Services.Configure<JwtConfiguration>(options => builder.Configuration.GetSection("Jwt").Bind(options));
 
+Console.WriteLine("checking URL Port: " + builder.Configuration["WEB_PORT"]);
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Program.cs:45]: Remove this

File: BackEnd/SnippetManagement.Api/Program.cs [hunk 26]
```diff
@@ -113,17 +117,14 @@
...
-
 app.UseAuthentication();
 app.UseAuthorization();
 
 app.MapControllers();
-app.UseFileServer( new FileServerOptions()
+app.UseFileServer(new FileServerOptions()
 {
-    FileProvider = new PhysicalFileProvider(Path.Combine(Directory.GetCurrentDirectory(),"Assets")),
+    FileProvider = new PhysicalFileProvider(Path.Combine(Directory.GetCurrentDirectory(), "Assets")),
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Api/Program.cs:127]: Assets as config value

File: BackEnd/SnippetManagement.DataModel/User.cs [hunk 27]
```diff
@@ -1,9 +1,12 @@
+using SnippetManagement.Common.Enum;
+
 namespace SnippetManagement.DataModel;
 
 public class User: BaseEntity<Guid>
```
Comment (minhman97) [inline BackEnd/SnippetManagement.DataModel/User.cs:5]: Space

File: BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs [hunk 28]
```diff
@@ -26,20 +29,42 @@ public AuthenticationService(IOptions<JwtConfiguration> jwtConfiguration, IUnitO
         _jwtConfiguration = jwtConfiguration.Value;
     }
 
-    public async Task<string> GetToken(UserDto userDto)
+    public async Task<Result<string>> GetToken(UserDto userDto, string externalToken)
```
Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs:32]: exToken nullable

Comment (minhman97) [inline BackEnd/SnippetManagement.Service/Services/Implementation/AuthenticationService.cs:32]: Refactor this, best to separate validation logic from business logic

File: FrontEnd/snippet_management.web/src/api/ApiEndpoint.js [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+export const programingLanguageUrl = `${baseUrl}/ProgramingLanguage`;
+
+export const getSnippetUrl = (
+  filterKeyWord,
+  startIndex,
+  endIndex,
+  sortProperty,
+  orderWay
+) => {
+  return filterKeyWord.trim() === ""
```
Comment (minhman97) [inline FrontEnd/snippet_management.web/src/api/ApiEndpoint.js:15]: Duplicate string

**Review by popemkt — APPROVED** [review]

**Pull request merged**